pool/src/pool.rs

```diff
@@ -0,0 +1,824 @@
+    blockchain: Arc<Box<DummyChain>>,
+}
+
+
+impl TransactionPool {
+    /// Searches for an output, designated by its commitment, from the current 
+    /// best UTXO view, presented by taking the best blockchain UTXO set (as
+    /// determined by the blockchain component) and rectifying pool spent and
+    /// unspents.
+    /// Detects double spends and unknown references from the pool and 
+    /// blockchain only; any conflicts with entries in the orphans set must
+    /// be accounted for separately, if relevant.
+    pub fn search_for_best_output(&self, output_commitment: &Commitment) -> Parent {
+        // The current best unspent set is: 
+        //   Pool unspent + (blockchain unspent - pool->blockchain spent)
+        // Pool unspents are unconditional so we check those first
+        self.pool.get_available_output(output_commitment).
+            map(|x| Parent::PoolTransaction{tx_ref: x.source_hash().unwrap()}).
+            or(self.search_blockchain_unspents(output_commitment)).
+            or(self.search_pool_spents(output_commitment)).
+            unwrap_or(Parent::Unknown)
+    }
+
+    // search_blockchain_unspents searches the current view of the blockchain
+    // unspent set, represented by blockchain unspents - pool spents, for an
+    // output designated by output_commitment.
+    fn search_blockchain_unspents(&self, output_commitment: &Commitment) -> Option<Parent> {
+        self.blockchain.get_best_utxo_set().get_output(output_commitment).
+            map(|o| match self.pool.get_blockchain_spent(output_commitment) {
+                Some(x) => Parent::AlreadySpent{other_tx: x.destination_hash().unwrap()},
+                None => Parent::BlockTransaction,
+            })
+    }
+
+    // search_pool_spents is the second half of pool input detection, after the
+    // available_outputs have been checked. This returns either a
+    // Parent::AlreadySpent or None.
+    fn search_pool_spents(&self, output_commitment: &Commitment) -> Option<Parent> {
+        self.pool.get_internal_spent(output_commitment).
+            map(|x| Parent::AlreadySpent{other_tx: x.destination_hash().unwrap()})
+
+    }
+
+    /// Get the number of transactions in the pool
+    pub fn pool_size(&self) -> usize {
+        self.pool.num_transactions()
+    }
+
+    pub fn orphans_size(&self) -> usize {
+        self.orphans.num_transactions()
+    }
+
+    pub fn total_size(&self) -> usize {
+        self.pool.num_transactions() + self.orphans.num_transactions()
+    }
+
+    /// Attempts to add a transaction to the pool.
+    ///
+    /// Adds a transation to the memory pool, deferring to the orphans pool
+    /// if necessary, and performing any connection-related validity checks.
+    /// Happens under an exclusive mutable reference gated by the write portion
+    /// of a RWLock.
+    ///
+    pub fn add_to_memory_pool(&mut self, source: TxSource, tx: transaction::Transaction) -> Result<(), PoolError> {
```

Comment:
Definitely agree, this makes sense to me. In a perfect world, I feel that orphans handling should occur under its own lock to avoid tying up pool resources for a lower-priority operation, which may help break this up as a consequence. In the meantime, I'll try to push more down into the Orphans container.